File: LudoGame/LudoGame/ViewModels/LudoViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class LudoViewModel: ObservableObject {
    @Published private(set) var state: LudoGameState
    @Published var displayedDiceValue: Int = 1
    @Published var isAnimatingDice = false

    private var engine: LudoGameEngine

    init(playerCount: Int = 4) {
        engine = LudoGameEngine(playerCount: playerCount)
        state = engine.state
    }

    func rollDice() {
        guard !isAnimatingDice,
              state.winner == nil,
              state.selectableTokenIDs.isEmpty else { return }

        isAnimatingDice = true
        Task {
            for _ in 0..<8 {
                displayedDiceValue = Int.random(in: 1...6)
                try? await Task.sleep(nanoseconds: 80_000_000)
            }
            engine.rollDice()
            if let roll = engine.state.lastDiceRoll {
                displayedDiceValue = roll
            }
            state = engine.state
            isAnimatingDice = false
        }
    }

    func selectToken(_ tokenID: Int) {
        engine.selectToken(tokenID)
        state = engine.state
    }

    func restart(playerCount: Int) {
        engine = LudoGameEngine(playerCount: playerCount)
        state = engine.state
        displayedDiceValue = 1
        isAnimatingDice = false
    }
}

File: LudoGameAndroid/app/src/main/java/com/ludogame/app/viewmodel/LudoViewModel.kt
package com.ludogame.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ludogame.app.engine.LudoGameEngine
import com.ludogame.app.models.LudoGameState
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.random.Random

class LudoViewModel(playerCount: Int) : ViewModel() {
    private var engine = LudoGameEngine(playerCount)

    private val _state = MutableStateFlow(engine.state)
    val state: StateFlow<LudoGameState> = _state.asStateFlow()

    private val _displayedDiceValue = MutableStateFlow(1)
    val displayedDiceValue: StateFlow<Int> = _displayedDiceValue.asStateFlow()

    private val _isAnimatingDice = MutableStateFlow(false)
    val isAnimatingDice: StateFlow<Boolean> = _isAnimatingDice.asStateFlow()

    fun rollDice() {
        val current = _state.value
        if (_isAnimatingDice.value || current.winner != null || current.selectableTokenIDs.isNotEmpty()) {
            return
        }

        viewModelScope.launch {
            _isAnimatingDice.value = true
            repeat(8) {
                _displayedDiceValue.value = Random.nextInt(1, 7)
                delay(80)
            }
            engine.rollDice()
            engine.state.lastDiceRoll?.let { _displayedDiceValue.value = it }
            _state.value = engine.state
            _isAnimatingDice.value = false
        }
    }

    fun selectToken(tokenId: Int) {
        engine.selectToken(tokenId)
        _state.value = engine.state
    }

    fun restart(playerCount: Int) {
        engine.restart(playerCount)
        _state.value = engine.state
        _displayedDiceValue.value = 1
        _isAnimatingDice.value = false
    }
}
